### DISCUSSION TITLE
getClassDependencies() and binary class interface

### DISCUSSION THREAD

**Discussion by loetifuss** [classgraph/classgraph#882] [Q&A]
Not sure if this is a bug or works as intended but I'm wondering about the behaviour of ClassGraph's ClassInfo.getClassDependencies() method.

Suppose I have the following setup:

An abstract parent Class with a private and a public method:

```
public abstract class ParentClass {

    private PrivateDependency privateDependency() {
        return new PrivateDependency();
    }

    public PublicDependency publicDependency() {
        return new PublicDependency();
    }
}
```

A class subclassing ParentClass:

```
public class SubClass extends ParentClass {
    public void subClassMethod() {
    }
}
```

Now, when I try to retrieve the class dependencies of `SubClass` ClassGraph returns both `PublicDependency` **and** `PrivateDependency` (see test below).
Since `privateDependency()` is a private method of the parent class, I would have expected that Classgraph returns only the dependencies "visible" from SubClass (sometimes referred to as 'application binary interface'). If this is the intended behaviour is there a way to configure ClassGraph to only return the dependencies visible from "SubClass"?

I have attached a small project reproducing the issue.
[classgraph-dependencies.zip](https://github.com/user-attachments/files/16942022/classgraph-dependencies.zip)

```
public class ClassGraphDependenciesTest {

    @Test
    void testClassGraphDependencies() {
        ClassGraph classGraph = new ClassGraph();

        classGraph.enableInterClassDependencies();
        classGraph.acceptPackages("org.example");

        try(ScanResult scan = classGraph.scan()) {
            ClassInfo classInfo = scan.getClassInfo("org.example.SubClass");
            ClassInfoList classDependencies = classInfo.getClassDependencies().directOnly();

            assertTrue(classDependencies.getNames().contains("org.example.PublicDependency"));
            assertFalse(classDependencies.getNames().contains("org.example.PrivateDependency"));
        }
    }
    
}
```

**Comment by lukehutch** [comment]
Hi, good catch... This API method was mostly designed to build visual dependency graphs, for people thinking about dependencies whether they were public or private. But probably it should respect whether the user has chosen to scan private methods/fields/classes or not.

Any chance you might want to try writing a PR for this?

**Reply by loetifuss** [reply]
Sure, I can give it a try.
